### ISSUE TITLE
Heap use-after-free in mrb_vm_exec

### ISSUE THREAD

**Issue by clayton-shopify** [mruby/mruby#3784]
The following input demonstrates a crash:
```ruby
class NoMethodError
  def initialize(message, name, args)
  ensure
    a super &name
  end

  b
end
```
It appears the problem was introduced in 58f5115b22ea0d4ac2869bb26d5b8830036c1aad.

ASAN report:
```ruby
==98305==ERROR: AddressSanitizer: heap-use-after-free on address 0x61e00000fa78 at pc 0x00010d34b578 bp 0x7fff5290bd70 sp 0x7fff5290bd68
READ of size 8 at 0x61e00000fa78 thread T0
    #0 0x10d34b577 in mrb_vm_exec (mruby:x86_64+0x100179577)
    #1 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #2 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #3 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #4 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #5 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #6 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #7 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #8 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #9 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #10 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #11 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #12 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #13 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #14 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #15 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #16 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #17 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #18 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #19 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #20 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #21 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #22 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #23 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #24 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #25 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #26 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #27 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #28 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #29 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #30 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #31 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #32 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #33 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #34 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #35 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #36 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #37 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #38 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #39 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #40 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #41 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #42 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #43 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #44 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #45 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #46 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #47 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #48 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #49 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #50 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #51 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #52 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #53 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #54 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #55 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #56 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #57 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #58 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #59 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #60 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #61 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #62 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #63 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #64 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #65 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #66 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #67 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #68 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #69 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #70 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #71 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #72 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #73 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #74 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #75 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #76 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #77 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #78 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #79 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #80 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #81 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #82 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #83 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #84 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #85 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #86 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #87 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #88 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #89 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #90 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #91 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #92 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #93 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #94 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #95 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #96 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #97 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #98 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #99 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #100 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #101 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #102 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #103 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #104 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #105 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #106 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #107 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #108 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #109 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #110 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #111 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #112 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #113 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #114 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #115 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #116 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #117 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #118 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #119 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #120 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #121 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #122 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #123 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #124 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #125 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #126 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #127 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #128 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #129 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #130 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #131 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #132 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #133 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #134 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #135 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #136 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #137 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #138 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #139 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #140 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #141 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #142 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #143 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #144 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #145 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #146 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #147 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #148 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #149 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #150 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #151 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #152 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #153 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #154 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #155 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #156 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #157 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #158 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #159 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #160 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #161 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #162 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #163 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #164 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #165 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #166 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #167 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #168 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #169 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #170 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #171 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #172 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #173 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #174 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #175 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #176 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #177 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #178 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #179 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #180 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #181 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #182 0x10d3722bf in mrb_top_run (mruby:x86_64+0x1001a02bf)
    #183 0x10d454b98 in mrb_load_exec (mruby:x86_64+0x100282b98)
    #184 0x10d4559c5 in mrb_load_file_cxt (mruby:x86_64+0x1002839c5)
    #185 0x10d1d48f3 in main mruby.c:227
    #186 0x7fffa6e81234 in start (libdyld.dylib:x86_64+0x5234)

0x61e00000fa78 is located 2552 bytes inside of 2560-byte region [0x61e00000f080,0x61e00000fa80)
freed by thread T0 here:
    #0 0x10d60b520 in wrap_realloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56520)
    #1 0x10d2cf535 in mrb_default_allocf (mruby:x86_64+0x1000fd535)
    #2 0x10d24ff28 in mrb_realloc_simple gc.c:204
    #3 0x10d25067e in mrb_realloc gc.c:218
    #4 0x10d333394 in cipush (mruby:x86_64+0x100161394)
    #5 0x10d33020b in mrb_funcall_with_block (mruby:x86_64+0x10015e20b)
    #6 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #7 0x10d2b25be in convert_type (mruby:x86_64+0x1000e05be)
    #8 0x10d2b2dbb in mrb_convert_type (mruby:x86_64+0x1000e0dbb)
    #9 0x10d34b35d in mrb_vm_exec (mruby:x86_64+0x10017935d)
    #10 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #11 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #12 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #13 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #14 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #15 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #16 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #17 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #18 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #19 0x10d243525 in mrb_no_method_error (mruby:x86_64+0x100071525)
    #20 0x10d27eb7e in mrb_method_missing (mruby:x86_64+0x1000acb7e)
    #21 0x10d346b8b in mrb_vm_exec (mruby:x86_64+0x100174b8b)
    #22 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #23 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #24 0x10d36f0b0 in ecall (mruby:x86_64+0x10019d0b0)
    #25 0x10d3455c5 in mrb_vm_exec (mruby:x86_64+0x1001735c5)
    #26 0x10d33c621 in mrb_vm_run (mruby:x86_64+0x10016a621)
    #27 0x10d333d6e in mrb_run (mruby:x86_64+0x100161d6e)
    #28 0x10d331834 in mrb_funcall_with_block (mruby:x86_64+0x10015f834)
    #29 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)

previously allocated by thread T0 here:
    #0 0x10d60b520 in wrap_realloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56520)
    #1 0x10d2cf535 in mrb_default_allocf (mruby:x86_64+0x1000fd535)
    #2 0x10d24ff28 in mrb_realloc_simple gc.c:204
    #3 0x10d25067e in mrb_realloc gc.c:218
    #4 0x10d251113 in mrb_malloc gc.c:240
    #5 0x10d2511ad in mrb_calloc gc.c:258
    #6 0x10d332265 in stack_init (mruby:x86_64+0x100160265)
    #7 0x10d32f2e0 in mrb_funcall_with_block (mruby:x86_64+0x10015d2e0)
    #8 0x10d32ec1a in mrb_funcall_with_block (mruby:x86_64+0x10015cc1a)
    #9 0x10d32e407 in mrb_funcall_argv (mruby:x86_64+0x10015c407)
    #10 0x10d216c75 in mrb_obj_new (mruby:x86_64+0x100044c75)
    #11 0x10d23adbd in mrb_exc_new_str (mruby:x86_64+0x100068dbd)
    #12 0x10d243e79 in mrb_init_exception (mruby:x86_64+0x100071e79)
    #13 0x10d274794 in mrb_init_core (mruby:x86_64+0x1000a2794)
    #14 0x10d2cf4ce in mrb_open_core (mruby:x86_64+0x1000fd4ce)
    #15 0x10d2cf69c in mrb_open_allocf (mruby:x86_64+0x1000fd69c)
    #16 0x10d2cf667 in mrb_open (mruby:x86_64+0x1000fd667)
    #17 0x10d1d3805 in main mruby.c:171
    #18 0x7fffa6e81234 in start (libdyld.dylib:x86_64+0x5234)

SUMMARY: AddressSanitizer: heap-use-after-free (mruby:x86_64+0x100179577) in mrb_vm_exec
Shadow bytes around the buggy address:
  0x1c3c00001ef0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3c00001f00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3c00001f10: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3c00001f20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3c00001f30: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x1c3c00001f40: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd[fd]
  0x1c3c00001f50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3c00001f60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3c00001f70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3c00001f80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3c00001f90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==98305==ABORTING
Abort trap: 6
```
This issue was reported by https://hackerone.com/flamezzz

**Comment by christopheraue** [comment]
I tried to reproduce this on Linux and Mac, but with no luck. Instead I get

```
-e:4:13: '&' interpreted as argument prefix
trace:
        [0] -e:1
        [1] -e:7
        [2] -e:4:in initialize
        [...]
        [256] -e:4:in initialize
-e:4: stack level too deep (SystemStackError)
```

**Comment by clayton-shopify** [comment]
@christopheraue Try will valgrind and you should see an invalid read:
```ruby
==1467== Invalid read of size 8
==1467==    at 0x42F01C: mrb_vm_exec (vm.c:1539)
==1467==    by 0x42C4F6: mrb_vm_run (vm.c:904)
==1467==    by 0x4357B0: mrb_run (vm.c:2850)
==1467==    by 0x42A8D8: ecall (vm.c:347)
==1467==    by 0x42DEE5: mrb_vm_exec (vm.c:1308)
==1467==    by 0x42C4F6: mrb_vm_run (vm.c:904)
==1467==    by 0x4357B0: mrb_run (vm.c:2850)
==1467==    by 0x42B33E: mrb_funcall_with_block (vm.c:480)
==1467==    by 0x42B3DD: mrb_funcall_argv (vm.c:490)
==1467==    by 0x417C4F: mrb_obj_new (class.c:1450)
==1467==    by 0x40D2FD: mrb_no_method_error (error.c:471)
==1467==    by 0x437BEF: mrb_method_missing (kernel.c:939)
==1467==  Address 0x5522c08 is 2,552 bytes inside a block of size 2,560 free'd
==1467==    at 0x4C2CE8E: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1467==    by 0x41A41E: mrb_default_allocf (state.c:60)
==1467==    by 0x4269BE: mrb_realloc_simple (gc.c:204)
==1467==    by 0x426A40: mrb_realloc (gc.c:218)
==1467==    by 0x42A44D: cipush (vm.c:264)
==1467==    by 0x42AF76: mrb_funcall_with_block (vm.c:434)
==1467==    by 0x42B3DD: mrb_funcall_argv (vm.c:490)
==1467==    by 0x4392B6: convert_type (object.c:320)
==1467==    by 0x4393B7: mrb_convert_type (object.c:342)
==1467==    by 0x42EFD1: mrb_vm_exec (vm.c:1536)
==1467==    by 0x42C4F6: mrb_vm_run (vm.c:904)
==1467==    by 0x4357B0: mrb_run (vm.c:2850)
==1467==
```
Or build with clang & ASAN: `CFLAGS=-fsanitize=address LDFLAGS=-fsanitize=address make`

**Comment by matz** [comment]
fixed by #3786
